File: Binary Tree Zigzag Level Order Traversal.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public List<List<Integer>> zigzagLevelOrder(TreeNode root) {
        List<List<Integer>> ans = new LinkedList<>() ;

        if (root == null) return ans ;

        Queue<TreeNode> q = new LinkedList<>() ;
        q.add(root) ;

        Integer flag = 0 ;

        while (q.isEmpty() == false) {
            Integer size = q.size() ;
            List<Integer> dummy = new LinkedList<>() ;
            while (size-- > 0) {
                TreeNode curr = q.poll() ;
                dummy.add(curr.val) ;
                if (curr.left != null) q.add(curr.left) ;
                if (curr.right != null) q.add(curr.right) ;
            }
            if (flag == 0) {
                ans.add(dummy) ;
            } else {
                Collections.reverse(dummy);
                ans.add(dummy) ;
            }
            flag = 1 - flag ;
        }

        return ans ;
    }
}

File: Binary Tree Zigzag Level Order Traversal.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    vector<vector<int>> zigzagLevelOrder(TreeNode* root) {
        vector<vector<int>> ans ;

        if (!root) return ans ;

        queue<TreeNode*> q ;

        q.push(root) ;

        int flag = 0 ;

        while (!q.empty()) {
            int size = q.size() ;
            vector<int> dummy ;
            while (size--) {
                TreeNode* curr = q.front() ; q.pop() ;
                dummy.push_back(curr->val) ;
                if (curr->left) q.push(curr->left) ;
                if (curr->right) q.push(curr->right) ;
            }
            if (flag == 0) {
                ans.push_back(dummy) ;
            }
            else {
                reverse(dummy.begin(), dummy.end()) ;
                ans.push_back(dummy) ;
            }
            flag = 1 - flag ;
        }
        return ans ;
    }
};